Added opentracing middleware

amarrella opened the pull request:
This is already a working version of the opentracing middleware.
I'm not sure on how to proceed wrt testing and if the approach I used is optimal.

Also, I think I would like to add the possibility to pass the span to the field context so that fields can pick it up and propagate the trace further.

OlegIlyenko commented on src/main/scala/sangria/slowlog/OpenTracing.scala:
```diff
@@ -0,0 +1,53 @@
+
+import scala.collection.concurrent.TrieMap
+
+class OpenTracing(implicit private val tracer: Tracer)
```
Also, do you think it would be a good idea to add another optional parameter for a parent `Span` for the whole thing? So that you can use it as a parent in `beforeQuery` if needed.

amarrella replied:
Adding it as optional parameter to the middleware would require recreating the middleware every time we receive a request. Not sure would be the best thing to do.

The approach i'm using in my project is "activate" the parent span just before calling `Executor.execute`, then the query span in `beforeQuery` by having nothing specified will pick up that span as parent.

OlegIlyenko commented on src/main/scala/sangria/slowlog/OpenTracing.scala:
```diff
@@ -0,0 +1,53 @@
+  type FieldVal = Unit
+
+  def beforeQuery(context: MiddlewareQueryContext[Any, _, _]) = {
+    val span = tracer.buildSpan(context.operationName.getOrElse("UNNAMED")).start()
```
What do you think about predicting the span name with `context.queryAst.operation(context.operationName).map(op ⇒ s"graphql-${op.operationType})` or something like this? Just to make it clear in the name that this span represent a GraphQL query/mutation/subscription

amarrella replied:
We could use tags for this as well. Tags would make it easier to filter and search particular spans. Need to find a nice key-value pair that fits well.
For example:
operation_type -> op.operationType
operation_type -> s"graphql-operationType"
graphql.operation_type -> op.operationType
What do you think?
For "UNNAMED" I like the idea of having it as parameter of `OpenTracing`.

amarrella commented:
@OlegIlyenko I added the unit tests and fixed a bug in the middleware. I also replied to a few of your comments and fixed others :) Can you have a look when you have time?
